### DISCUSSION TITLE
Spark Splink performance metrics

### DISCUSSION THREAD

**Discussion by hadoan88** [moj-analytical-services/splink#2245] [General]
I'm a newbie with Splink. I want to discover Spark Splink performance metrics on 100M golden records. 
 What is the min/max time value to find a matching record?  I use **first_name, last_name, email, phone, and address** to dedup the record

Does anyone have those metrics? 

Thanks in advance.

**Comment by vfrank66** [comment]
Performance is dependent on the comparisons columns as you stated + plus the data, and also the blocking rules. Generally if you have the total number of comparisons that is better information to determine time performance. 

Combing these two details plus the state of your data will inform you of the results you are looking. In my model which has similar comparison levels plus a few more I can predict on 19 mil records and cluster those predictions, then save them to s3 in 30 minutes. The slowness for me is the .as_pandas_dataframe(). But again my blocking if different and my data is different.

**Reply by RobinL** [reply]
You may be already aware, but you can use `as_spark_dataframe` unless you really need the pandas data frame

**Reply by vfrank66** [reply]
Haha no sir I did not catch that originally, I see it now in the source code. I based my spark off the doc's which only used as_pandas_dataframe() so good to know. That would greatly reduce the 30 minutes.

**Reply by hadoan88** [reply]
Thanks @vfrank66 , @RobinL, basically I would like to explore the performance of both functions **find_matches_to_new_records** and **compare_two_records** when a new record comes in and do a matching.

I tested the execution time of this example https://moj-analytical-services.github.io/splink/demos/examples/spark/deduplicate_1k_synthetic.html on a Databricks spark cluster - sing node (14GB 4 cores)
1. _find_matches_to_new_records_ - took less than **3s** to respond
2. _compare_two_records_ - took around **8s** to respond

As above result, I wonder why _compare_two_records_ took more time than _find_matches_to_new_records_


Is it possible to respond in less than **5 minutes** for _find_matches_to_new_records_ and _compare_two_records_ with a dataset of 100M records in Spark?

**Reply by RobinL** [reply]
> Haha no sir I did not catch that originally, I see it now in the source code. I based my spark off the doc's which only used as_pandas_dataframe() so good to know. That would greatly reduce the 30 minutes.

That's a good point, i'll try to remember to update that example in the docs to use the method.  You're not the first to miss it!  Edit: [done in splink4](https://github.com/moj-analytical-services/splink/pull/2247)

**Reply by hadoan88** [reply]
Hi @RobinL, I tested **find_matches_to_new_records** method and got a different execution time since the blocking condition is the same

`new_record_df = pd.DataFrame(
        {
            "unique_id": [90000],
            "first_name": ['chloe'],
            "surname": ['santos'],
            "dob": ['2022-01-25'],
            "city": ['Ho Chi Minh city'],
            "email": ['claudia65@huynh.com'],
        }
    )

> record_predictions = (linker.find_matches_to_new_records(new_record_df, blocking_rules=[**block_on("first_name")**, block_on("surname")])

--> took _3(s)_ to complete

> record_predictions = (linker.find_matches_to_new_records(new_record_df, blocking_rules=[**block_on("l.first_name = r.surname")**, block_on("surname")])

--> took _4 (mins)_ to complete

Are there any magic behind?

**Reply by vfrank66** [reply]
@hadoan88 your blocking conditions are not the same therefore you are creating different comparisons. Luckily it is not magical, but to help quantify your research you should read through and test this very good documentation: https://moj-analytical-services.github.io/splink/demos/tutorials/03_Blocking.html?h=count#counting-the-number-of-comparisons-created-by-a-list-of-blocking-rules

**Reply by hadoan88** [reply]
Thanks @vfrank66, it's weird. In my mind, I think **block_on("first_name")** is equivalent to **block_on("l.first_name = r.surname")**

**Reply by RobinL** [reply]
block_on("l.first_name = r.surname") will produce strange results. The correct syntax for this rule is just "l.first_name = r.surname" i.e. a string
